todo-list-backend/src/index.ts

```diff
@@ -0,0 +1,124 @@
+        updateTodoDone(id: Int!, done: Boolean!): ToDo
+        deleteTodo(id: Int!): ToDo
+        createUser(name: String! password: String!): User
+    }
+`;
+
+function buildResolvers(storage: IStorage, authSecret: string): any {
+    return {
+        Query: {
+            readTodos: (parent: any, args: any, context: any, info: any) => {
+                let user: User = context.user;
+                return storage.readTodos(user);
+            },
+        },
+
+        Mutation: {
+            login: (parent: any, args: any, context: any, info: any) => {
+                let user = storage.readUser(args.name);
+
+                if (!user || !user.checkPassword(args.password)) {
+                    return null;
+                }
+
+                let payload = {
+                    username: user.name
+                };
+
+                return sign(payload, authSecret, {expiresIn: "1 day"});
+            },
+            createTodo: (parent: any, args: any, context: any, info: any) => {
+                let user: User = context.user;
+                return storage.createTodo(user, args.name);
+            },
+            updateTodoName: (parent: any, args: any, context: any, info: any) => {
+                return storage.updateTodoName(context.user, args.id, args.name);
+            },
+            updateTodoDone: (parent: any, args: any, context: any, info: any) => {
+                return storage.updateTodoDone(context.user, args.id, args.done);
+            },
+            deleteTodo: (parent: any, args: any, context: any, info: any) => {
+                return storage.deleteTodo(context.user, args.id);
+            },
+            createUser: (parent: any, args: any, context: any, info: any) => {
+                return storage.createUser(args.name, args.password);
+            }
+        }
+    }
+}
+
+export function buildApolloServer(storage: IStorage, authSecret: string,
+                                  debugMode: boolean = false, context?: any): ApolloServer {
+
+    const resolvers = buildResolvers(storage, authSecret);
+    const schema = makeExecutableSchema({typeDefs, resolvers});
+    const permissions = buildPermissions(schema, authSecret, debugMode);
+    const finalizedSchema = applyMiddleware(schema, permissions);
+
+    return new ApolloServer({
+        schema: finalizedSchema,
+        context: ({req}) => {
+            if (context) {
+                return context;
+            }
+
+            let token: string = req.headers.authorization || "";
+            let user: User = null;
+            let decodedToken: any = decode(token, {complete: true});
+            if (decodedToken) {
+                let username = decodedToken.payload.username;
+                user = storage.readUser(username);
+            }
+
+            return {
+                token: token,
+                user: user,
+            }
+        }
+    });
+}
+
+if (require.main === module) {
```

Comment:
Ah, nice! I've never seen that in JavaScript! Learned something new today :tada: Reminds me of the same idea in python.

At Human Connection we store our secrets in a `.gitignore`d file `.env` and we read those variables with [dotenv](https://github.com/motdotla/dotenv).